go/runtime/host/protocol/connection.go

```diff
@@ -0,0 +1,461 @@
+package protocol
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"sync"
+
+	"github.com/opentracing/opentracing-go"
+	opentracingExt "github.com/opentracing/opentracing-go/ext"
+
+	"github.com/oasislabs/oasis-core/go/common"
+	"github.com/oasislabs/oasis-core/go/common/cbor"
+	"github.com/oasislabs/oasis-core/go/common/errors"
+	"github.com/oasislabs/oasis-core/go/common/logging"
+	"github.com/oasislabs/oasis-core/go/common/tracing"
+	"github.com/oasislabs/oasis-core/go/common/version"
+)
+
+const moduleName = "rhp/internal"
+
+var (
+	// ErrNotReady is the error reported when the Runtime Host Protocol is not initialized.
+	ErrNotReady = errors.New(moduleName, 1, "rhp: not ready")
+)
+
+// Handler is a protocol message handler interface.
+type Handler interface {
+	// Handle given request and return a response.
+	Handle(ctx context.Context, body *Body) (*Body, error)
+}
+
+// Connection is a Runtime Host Protocol connection interface.
+type Connection interface {
+	// Close closes the connection.
+	Close()
+
+	// Call sends a request to the other side and returns the response or error.
+	Call(ctx context.Context, body *Body) (*Body, error)
+
+	// InitHost performs initialization in host mode and transitions the connection to Ready state.
+	//
+	// This method must be called before the host will answer requests.
+	//
+	// Only one of InitHost/InitGuest can be called otherwise the method may panic.
+	//
+	// Returns the self-reported runtime version.
+	InitHost(ctx context.Context, conn net.Conn) (*version.Version, error)
+
+	// InitGuest performs initialization in guest mode and transitions the connection to Ready
+	// state.
+	//
+	// Only one of InitHost/InitGuest can be called otherwise the method may panic.
+	InitGuest(ctx context.Context, conn net.Conn) error
+}
+
+// state is the connection state.
+type state uint8
+
+const (
+	stateUninitialized state = iota
+	stateInitializing
+	stateReady
+	stateClosed
+)
+
+func (s state) String() string {
+	switch s {
+	case stateUninitialized:
+		return "uninitialized"
+	case stateInitializing:
+		return "initializing"
+	case stateReady:
+		return "ready"
+	case stateClosed:
+		return "closed"
+	default:
+		return fmt.Sprintf("[malformed: %d]", s)
+	}
+}
+
+type connection struct { // nolint: maligned
+	sync.RWMutex
+
+	conn  net.Conn
+	codec *cbor.MessageCodec
+
+	runtimeID common.Namespace
+	handler   Handler
+
+	state           state
+	pendingRequests map[uint64]chan *Body
+	nextRequestID   uint64
+
+	outCh   chan *Message
+	closeCh chan struct{}
+	quitWg  sync.WaitGroup
+
+	logger *logging.Logger
+}
+
+func (c *connection) getState() state {
+	c.RLock()
+	s := c.state
+	c.RUnlock()
+	return s
+}
+
+func (c *connection) setStateLocked(state state) {
+	// TODO: Validate state transition.
+	c.state = state
+}
+
+// Implements Connection.
+func (c *connection) Close() {
+	if c.getState() != stateReady {
+		return
+	}
+
+	if err := c.conn.Close(); err != nil {
```

Comment:
Isn't this race condition prone?  `Close` and `Call` both dereference `c.conn`, and the critical section is too small.